{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Models\n",
    "======\n",
    "\n",
    "The starting point for these methods is a mixed linear model of the form:\n",
    "\n",
    "\n",
    "  $$\\boldsymbol{y=X\\beta+Z\\alpha+e}$$\n",
    "\n",
    "where  $y$ is an $n\\times 1$ vector of trait phenotypic values, $\\boldsymbol{X}$ is an $n\\times p$ incidence matrix relating the vector $\\boldsymbol{\\beta}$ of non-genetic fixed or random effects to $\\boldsymbol{y}$,  $\\boldsymbol{Z}$\n",
    "is an $n\\times k$ matrix of genotype covariates (coded as 0, 1 or 2)\n",
    "for $k$ SNP markers, $\\boldsymbol{\\alpha}$ is a $k\\times 1$ vector of random\n",
    "partial regression coefficients of the $k$ SNPs (which are more\n",
    "commonly referred to as the marker effects), and $\\boldsymbol{e}$ is a\n",
    "vector of residuals. \n",
    "\n",
    "To proceed with Bayesian regression, prior distributions must be\n",
    "specified for $\\beta$, $\\alpha$ and $e$. In all the models\n",
    "considered here a flat prior is used for\n",
    "$\\boldsymbol{\\beta}$, and conditional on the residual variance, $\\sigma^2_e$, a\n",
    "normal distribution with null mean and covariance matrix\n",
    "$\\sigma^2_e$ is used for the vector of residuals, where $R$\n",
    "is a diagonal matrix. Further, $\\sigma^2_e$ is treated as an unknown\n",
    "with a scaled inverse chi-square prior. The alternative methods differ \n",
    "only in the prior used for $\\alpha$.\n",
    "\n",
    "### BayesA\n",
    "\n",
    "The prior assumption is that marker effects have identical\n",
    "and independent univariate-t distributions each with a null mean,\n",
    "scale parameter $S^2_{\\alpha}$ and $\\nu$ degrees of freedom.\n",
    "This is equivalent to assuming that the marker effect at locus $i$ has a univariate normal\n",
    "with null mean and unknown, locus-specific variance :math:$\\sigma^2_i$,\n",
    "which in turn is assigned a scaled inverse chi-square prior with scale\n",
    "parameter $S^2_{\\alpha}$ and $\\nu_{\\alpha}$ degrees of freedom. \n",
    "\n",
    "### BayesB\n",
    "\n",
    "In BayesB, the prior assumption is that marker effects have identical\n",
    "and independent mixture distributions, where each has a point mass at\n",
    "zero with probability $\\pi$ and a univariate-t distribution with\n",
    "probability $1-\\pi$ having a null mean, scale parameter $S^2_{\\alpha}$\n",
    "and $\\nu$ degrees of freedom. Thus, BayesA is a special case of BayesB\n",
    "with $\\pi=0$. Further, as in BayesA, the t-distribution in BayesB is\n",
    "equivalent to a univariate normal with null mean and unknown,\n",
    "locus-specific variance, which in turn is assigned a scaled inverse chi-square\n",
    "prior with scale parameter $S^2_{\\alpha}$ and $\\nu_{\\alpha}$ degrees\n",
    "of freedom. \n",
    "\n",
    "Here, we introduce a third form of the BayesB prior that is equivalent\n",
    "to the two given above in that all three produce the same posterior\n",
    "for locus effects. To do so, we introduce a Bernoulli variable\n",
    "$\\delta_i$ for locus $i$ that is 1 with probability $1-\\pi$ and zero\n",
    "with probability $\\pi$. Then, the effect of locus $i$ is written as\n",
    "\n",
    "  $$\\alpha_i = \\zeta_i\\delta_i$$\n",
    "\n",
    "where $\\zeta_i$ has a normal distribution with null mean and\n",
    "locus-specific variance $\\sigma^2_i$, which in turn has a scaled\n",
    "inverse chi-square prior with scale parameter $S^2_{\\alpha}$ and\n",
    "$\\nu_{\\alpha}$ degrees of freedom.\n",
    "\n",
    "\n",
    "### BayesC and BayesCpi\n",
    "\n",
    "In BayesC, the prior assumption is that marker effects have identical\n",
    "and independent mixture distributions, where each has a point mass at\n",
    "zero with probability $\\pi$ and a univariate-normal distribution with\n",
    "probability $1-\\pi$ having a null mean and variance\n",
    "$\\sigma^2_{\\alpha}$, which in turn has a scaled inverse chi-square\n",
    "prior with scale parameter $S^2_{\\alpha}$ and $\\nu_{\\alpha}$ degrees\n",
    "of freedom.  \n",
    "In addition to the above assumptions, in BayesC $\\pi$, $\\pi$ is treated\n",
    "as unknown with a uniform prior. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### reference\n",
    "* Fernando RL, Garrick D. Bayesian methods applied to GWAS. Methods Mol Biol. 2013;1019:237â€“274. doi: 10.1007/978-1-62703-447-0_10"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 0.4.5",
   "language": "julia",
   "name": "julia-0.4"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "0.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
