<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Public · JWAS.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../index.html"><img class="logo" src="../../assets/logo.png" alt="JWAS.jl logo"/></a><h1>JWAS.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><a class="toctext" href="../../theory/theory/">Some Theory</a></li><li><a class="toctext" href="../../citing/citing/">Citing</a></li><li><span class="toctext">Manual</span><ul><li><a class="toctext" href="../getstarted/">Get Started</a></li><li><a class="toctext" href="../workflow/">Workflow</a></li><li class="current"><a class="toctext" href>Public</a><ul class="internal"><li><a class="toctext" href="#Index-1">Index</a></li><li><a class="toctext" href="#Public-Interface-1">Public Interface</a></li></ul></li><li><a class="toctext" href="../internals/">Internals</a></li></ul></li><li><span class="toctext">Examples</span><ul><li><a class="toctext" href="../../examples/examples/">Examples</a></li></ul></li><li><a class="toctext" href="../../FrequentlyAskedQuestions/FrequentlyAskedQuestions/">Frequently Asked Questions</a></li></ul></nav><article id="docs"><header><nav><ul><li>Manual</li><li><a href>Public</a></li></ul><a class="edit-page" href="https://github.com/reworkhow/JWAS.jl/blob/master/docs/src/manual/public.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Public</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Public-functions-1" href="#Public-functions-1">Public functions</a></h1><p>Documentation for JWAS.jl&#39;s public interface. Below are functions available to general users.</p><h2><a class="nav-anchor" id="Index-1" href="#Index-1">Index</a></h2><ul><li><a href="#JWAS.add_genotypes"><code>JWAS.add_genotypes</code></a></li><li><a href="#JWAS.build_model"><code>JWAS.build_model</code></a></li><li><a href="#JWAS.outputEBV"><code>JWAS.outputEBV</code></a></li><li><a href="#JWAS.outputMCMCsamples"><code>JWAS.outputMCMCsamples</code></a></li><li><a href="#JWAS.runMCMC"><code>JWAS.runMCMC</code></a></li><li><a href="#JWAS.set_covariate"><code>JWAS.set_covariate</code></a></li><li><a href="#JWAS.set_random"><code>JWAS.set_random</code></a></li><li><a href="#JWAS.showMME"><code>JWAS.showMME</code></a></li><li><a href="#JWAS.solve"><code>JWAS.solve</code></a></li></ul><ul><li><a href="#JWAS.misc.GWAS"><code>JWAS.misc.GWAS</code></a></li><li><a href="#JWAS.misc.QC"><code>JWAS.misc.QC</code></a></li><li><a href="#JWAS.misc.get_additive_genetic_variances"><code>JWAS.misc.get_additive_genetic_variances</code></a></li><li><a href="#JWAS.misc.get_breeding_values"><code>JWAS.misc.get_breeding_values</code></a></li><li><a href="#JWAS.misc.get_correlations"><code>JWAS.misc.get_correlations</code></a></li><li><a href="#JWAS.misc.get_heritability"><code>JWAS.misc.get_heritability</code></a></li></ul><h2><a class="nav-anchor" id="Public-Interface-1" href="#Public-Interface-1">Public Interface</a></h2><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="JWAS.build_model" href="#JWAS.build_model"><code>JWAS.build_model</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">build_model(model_equations::AbstractString,R;df::Float64=4.0)</code></pre><ul><li>Build models from <strong>model equations</strong> with residual varainces <strong>R</strong> and degree of freedom for residual variance <strong>df</strong> defaulting to 4.0.</li><li>By default, all variabels in model<em>equations are fixed and factors. Set variables to be covariates or random using functions `set</em>covariate()<code>or</code>set_random()`.</li></ul><pre><code class="language-julia">#single-trait
model_equations = &quot;BW = intercept + age + sex&quot;
R               = 6.72
models          = build_model(model_equations,R);

#multi-trait
model_equations = &quot;BW = intercept + age + sex
                   CW = intercept + litter&quot;;
R               = [6.72   24.84
                   24.84  708.41]
models          = build_model(model_equations,R);</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/1.JWAS/src/build_MME.jl#L14-L35">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="JWAS.set_covariate" href="#JWAS.set_covariate"><code>JWAS.set_covariate</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">set_covariate(mme::MME,variables::AbstractString...)</code></pre><ul><li>set <strong>variables</strong> as covariates; <strong>mme</strong> is the output of function <code>build_model()</code>.</li></ul><pre><code class="language-julia">#After running build_model, variabels age and year can be set to be covariates as
set_covariate(models,&quot;age&quot;,&quot;year&quot;)
#or
set_covariate(models,&quot;age year&quot;)</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/1.JWAS/src/build_MME.jl#L64-L75">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="JWAS.set_random" href="#JWAS.set_random"><code>JWAS.set_random</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">set_random(mme::MME,randomStr::AbstractString,ped::Pedigree, G;df=4)</code></pre><ul><li>set variables as random polygenic effects with pedigree information <strong>ped</strong> and variances <strong>G</strong> with degree of freedom <strong>df</strong>, defaulting to <code>4.0</code>.</li></ul><pre><code class="language-julia">#single-trait (example 1)
model_equation  = &quot;y = intercept + age + animal&quot;
model           = build_model(model_equation,R)
ped             = get_pedigree(pedfile)
G               = 1.6
set_random(model,&quot;animal&quot;, ped, G)

#single-trait (example 2)
model_equation  = &quot;y = intercept + age + animal + animal*age&quot;
model           = build_model(model_equation,R)
ped             = get_pedigree(pedfile)
G               = [1.6   0.2
                   0.2  1.0]
set_random(model,&quot;animal animal*age&quot;, ped,G)

#multi-trait
model_equations = &quot;BW = intercept + age + sex + animal
                   CW = intercept + age + sex + animal&quot;
model           = build_model(model_equations,R);
ped             = get_pedigree(pedfile);
G               = [6.72   2.84
                   2.84  8.41]
set_random(model,&quot;animal&quot;,ped,G)</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/1.JWAS/src/random_effects.jl#L5-L36">source</a><div><div><pre><code class="language-none">set_random(mme::MME,randomStr::AbstractString,G;Vinv=0,names=[],df=4)</code></pre><ul><li>set variables as random effects, defaulting to i.i.d effects, with variances <strong>G</strong> with degree of freedom <strong>df</strong>, defaulting to 4.0.</li><li>the random effects are assumed to be i.i.d by default and it can be defined with any (inverse of) covariance structure <strong>Vinv</strong> with its index (row names) provided by <strong>names</strong>.</li></ul><pre><code class="language-julia">#single-trait (i.i.d randome effects)
model_equation  = &quot;y = intercept + litter + sex&quot;
model           = build_model(model_equation,R)
G               = 0.6
set_random(model,&quot;litter&quot;,G)

#multi-trait (i.i.d randome effects)
model_equations = &quot;BW = intercept + litter + sex
                   CW = intercept + litter + sex&quot;
model           = build_model(model_equations,R);
G               = [3.72  1.84
                   1.84  3.41]
set_random(model,&quot;litter&quot;,G)

#single-trait (randome effects with specific covariance structures)
model_equation  = &quot;y = intercept + litter + sex&quot;
model           = build_model(model_equation,R)
V               = [1.0  0.5 0.25
                   0.5  1.0 0.5
                   0.25 0.5 1.0]
G               = 0.6
set_random(model,&quot;litter&quot;,G,Vinv=inv(V),names=[a1;a2;a3])</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/1.JWAS/src/random_effects.jl#L84-L116">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="JWAS.PedModule.get_pedigree" href="#JWAS.PedModule.get_pedigree"><code>JWAS.PedModule.get_pedigree</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">get_pedigree(pedfile::AbstractString;header=false,separator=&#39;,&#39;)</code></pre><ul><li>Get pedigree informtion from a pedigree file with <strong>header</strong> (defaulting to <code>false</code>) and <strong>separator</strong> (defaulting to <code>,</code>).</li><li>Pedigree file format:</li></ul><pre><code class="language-none">a,0,0
c,a,b
d,a,c</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/2.PedModule/src/PedModule.jl#L10-L21">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="JWAS.add_genotypes" href="#JWAS.add_genotypes"><code>JWAS.add_genotypes</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">add_genotypes(mme::MME,M,G;header=false,center=true,rowID=false,G_is_marker_variance=false,df=4)</code></pre><ul><li>Get marker informtion from an nxp Matrix M of genotypes (Array or DataFrame), where n is the number of individuals and p is the number of markers. This matrix needs to be column-wise sorted by marker positions.</li><li>rowID is a vector of individual IDs; if it is omitted, IDs will be set to 1:n</li><li>header is a header vector such as [&quot;id&quot;; &quot;mrk1&quot;; &quot;mrk2&quot;;...;&quot;mrkp&quot;]. If omitted, marker names will be set to 1:p</li><li><strong>G</strong> defaults to the genetic variance with degree of freedom <strong>df</strong>=4.0.</li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/1.JWAS/src/markers/readgenotypes.jl#L1-L9">source</a><div><div><pre><code class="language-none">add_genotypes(mme::MME,file,G;separator=&#39; &#39;,header=true,center=true,G_is_marker_variance=false,df=4.0)</code></pre><ul><li>Get marker informtion from a genotype file. This file needs to be column-wise sorted by marker positions.</li><li><strong>G</strong> defaults to the genetic variance with degree of freedom <strong>df</strong>=4.0.</li><li>File format:</li></ul><pre><code class="language-none">Animal,marker1,marker2,marker3,marker4,marker5
S1,1,0,1,1,1
D1,2,0,2,2,1
O1,1,2,0,1,0
O3,0,0,2,1,1</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/1.JWAS/src/markers/readgenotypes.jl#L28-L41">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="JWAS.runMCMC" href="#JWAS.runMCMC"><code>JWAS.runMCMC</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">runMCMC(model::MME,df::DataFrame;
        ### MCMC
        chain_length             = 1000,
        starting_value           = false,
        burnin                   = 0,
        output_samples_frequency = 0,
        output_samples_file      = &quot;MCMC_samples&quot;,
        update_priors_frequency  = 0,
        ### Methods
        methods                  = &quot;conventional (no markers)&quot;,
        estimate_variance        = true,
        Pi                       = 0.0,
        estimatePi               = false,
        estimateScale            = false,
        single_step_analysis     = false,
        pedigree                 = false,
        categorical_trait        = false,
        missing_phenotypes       = false,
        constraint               = false,
        causal_structure         = false,
        ### Genomic Prediction
        outputEBV                = true,
        output_heritability      = false,
        ### MISC
        seed                     = false,
        printout_model_info      = true,
        printout_frequency       = 0)</code></pre><p><strong>Run MCMC for Bayesian Linear Mixed Models with or without estimation of variance components.</strong></p><ul><li>Markov chain Monte Carlo<ul><li>The first <code>burnin</code> iterations are discarded at the beginning of a MCMC chain of length <code>chain_length</code>.</li><li>The <code>starting_value</code> can be provided as a vector for all location parameteres and marker effects, defaulting to <code>0.0</code>s.</li><li>Save MCMC samples every <code>output_samples_frequency</code> iterations, defaulting to <code>chain_length/1000</code>, to files <code>output_samples_file</code>, defaulting to <code>MCMC_samples.txt</code>. MCMC samples for hyperparametes (variance componets) and marker effects are saved by default. MCMC samples for location parametes can be saved using <code>output_MCMC_samples()</code>. Note that saving MCMC samples too frequently slows down the computation.</li><li>Miscellaneous Options<ul><li>Priors are updated every <code>update_priors_frequency</code> iterations, defaulting to <code>0</code>.</li></ul></li></ul></li><li>Methods<ul><li>Available <code>methods</code> include &quot;conventional (no markers)&quot;, &quot;RR-BLUP&quot;, &quot;BayesB&quot;, &quot;BayesC&quot;, &quot;Bayesian Lasso&quot;, and &quot;GBLUP&quot;.</li><li>Single step analysis is allowed if <code>single_step_analysis</code> = <code>true</code> and <code>pedigree</code> is provided.</li><li>In Bayesian variable selection methods, <code>Pi</code> for single-trait analyses is a number; <code>Pi</code> for multi-trait analyses is a dictionary such as <code>Pi=Dict([1.0; 1.0]=&gt;0.7,[1.0; 0.0]=&gt;0.1,[0.0; 1.0]=&gt;0.1,[0.0; 0.0]=&gt;0.1)</code>, defaulting to <code>all markers have effects (Pi = 0.0)</code> in single-trait analysis and <code>all markers have effects on all traits (Pi=Dict([1.0; 1.0]=&gt;1.0,[0.0; 0.0]=&gt;0.0))</code> in multi-trait analysis. <code>Pi</code> is estimated if <code>estimatePi</code> = true</li><li>Variance components are estimated if <code>estimate_variance</code>=true, defaulting to <code>true</code>.</li><li>Scale parameter for prior of marker effect variance is estimated if <code>estimateScale</code> = true</li><li>Miscellaneous Options<ul><li>Missing phenotypes are allowed in multi-trait analysis with <code>missing_phenotypes</code>=true, defaulting to <code>true</code>.</li><li>Catogorical Traits are allowed if <code>categorical_trait</code>=true, defaulting to <code>false</code>.</li><li>If <code>constraint</code>=true, defaulting to <code>false</code>, constrain residual covariances between traits to be zeros.</li><li>If <code>causal_structure</code> is provided, e.g., causal_structure = [0.0,0.0,0.0;1.0,0.0,0.0;1.0,0.0,0.0] for trait 2 -&gt; trait 1 and trait 3 -&gt; trait 1, phenotypic causal networks will be incorporated using structure equation models.</li></ul></li></ul></li><li>Genomic Prediction<ul><li>Individual estimted breeding values (EBVs) and prediction error variances (PEVs) are returned if <code>outputEBV</code>=true, defaulting to <code>true</code>. Heritability and genetic</li></ul>variances are returned if <code>output_heritability</code>=<code>true</code>, defaulting to <code>false</code>. Note that estimation of heritability is computaionally intensive.</li><li>Miscellaneous Options<ul><li>Print out the model information in REPL if <code>printout_model_info</code>=true; print out the monte carlo mean in REPL with <code>printout_frequency</code>, defaulting to <code>false</code>.</li><li>If <code>seed</code>, defaulting to <code>false</code>, is provided, a reproducible sequence of numbers will be generated for random number generation.</li></ul></li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/1.JWAS/src/MCMC/MCMC.jl#L10-L72">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="JWAS.outputEBV" href="#JWAS.outputEBV"><code>JWAS.outputEBV</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">outputEBV(model,IDs::Array;PEV=false)</code></pre><p>Output estimated breeding values and prediction error variances (defaulting to false) for IDs.</p></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/1.JWAS/src/MCMC/output.jl#L119-L123">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="JWAS.outputMCMCsamples" href="#JWAS.outputMCMCsamples"><code>JWAS.outputMCMCsamples</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">outputMCMCsamples(mme::MME,trmStr::AbstractString...)</code></pre><p>Get MCMC samples for specific location parameters.</p></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/1.JWAS/src/MCMC/outputMCMCsamples.jl#L5-L9">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="JWAS.showMME" href="#JWAS.showMME"><code>JWAS.showMME</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">showMME(mme::MME,df::DataFrame)</code></pre><ul><li>Show left-hand side and right-hand side of mixed model equations (no markers).</li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/1.JWAS/src/printout.jl#L25-L29">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="JWAS.solve" href="#JWAS.solve"><code>JWAS.solve</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">solve(mme::MME,df::DataFrame;solver=&quot;default&quot;,printout_frequency=100,tolerance = 0.000001,maxiter = 5000)</code></pre><ul><li>Solve the mixed model equations (no marker information) without estimating variance components.</li></ul><p>Available solvers include <code>default</code>, <code>Jacobi</code>, <code>GaussSeidel</code>, and <code>Gibbs sampler</code>.</p></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/1.JWAS/src/solver.jl#L1-L6">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="JWAS.misc.GWAS" href="#JWAS.misc.GWAS"><code>JWAS.misc.GWAS</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">GWAS(marker_effects_file;header=false)</code></pre><p>Compute the model frequency for each marker (the probability the marker is included in the model) using samples of marker effects stored in <strong>marker<em>effects</em>file</strong>.</p></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/4.misc/src/GWAS.jl#L1-L5">source</a><div><div><pre><code class="language-none">GWAS(marker_effects_file,model;header=true,window_size=&quot;1 Mb&quot;,threshold=0.001)</code></pre><p>run genomic window-based GWAS without marker locations</p><ul><li>MCMC samples of marker effects are stored in <strong>marker<em>effects</em>file</strong> with delimiter &#39;,&#39;.</li><li><strong>window_size</strong> is either a constant (identical number of markers in each window) or an array of number of markers in each window</li><li><strong>model</strong> is either the model::MME used in analysis or the genotypic covariate matrix M::Array</li><li>File format:</li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/4.misc/src/GWAS.jl#L23-L33">source</a><div><div><pre><code class="language-none">GWAS(marker_effects_file,map_file,model;header=true,window_size=&quot;1 Mb&quot;,threshold=0.001)</code></pre><p>run genomic window-based GWAS</p><ul><li>MCMC samples of marker effects are stored in <strong>marker<em>effects</em>file</strong> with delimiter &#39;,&#39;.</li><li><strong>model</strong> is either the model::MME used in analysis or the genotypic cavariate matrix M::Array</li><li><strong>map_file</strong> has the (sorted) marker position information with delimiter &#39;,&#39;.</li><li>File format:</li></ul><pre><code class="language-none">markerID,chromosome,position
m1,1,16977
m2,1,434311
m3,1,1025513
m4,2,70350
m5,2,101135</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/4.misc/src/GWAS.jl#L82-L102">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="JWAS.misc.get_additive_genetic_variances" href="#JWAS.misc.get_additive_genetic_variances"><code>JWAS.misc.get_additive_genetic_variances</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">get_additive_genetic_variances(model::MME,files...;header=true)</code></pre><ul><li>Get MCMC samples for additive genetic variances using samples for marker effects stored in <strong>files</strong>.</li><li>Return a vector for single-trait analysis and an array of matrices for multi-trait analysis</li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/4.misc/src/using_marker_samples.jl#L122-L127">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="JWAS.misc.get_heritability" href="#JWAS.misc.get_heritability"><code>JWAS.misc.get_heritability</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">get_heritability(samples_for_genetic_variances::Array{Array{Float64,2},1},samples_for_residual_variances::Array{Array{Float64,2},1}))</code></pre><ul><li>Get MCMC samples for heritabilities using MCMC samples for genetic variances and residual variances for multi-trait analysis</li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/4.misc/src/using_marker_samples.jl#L132-L135">source</a><div><div><pre><code class="language-none">get_heritability(samples_for_genetic_variances::Array{Float64,1},samples_for_residual_variances::Array{Float64,1}))</code></pre><ul><li>Get MCMC samples for heritabilities using MCMC samples for genetic variances and residual variances for single-trait analysis</li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/4.misc/src/using_marker_samples.jl#L147-L150">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="JWAS.misc.get_correlations" href="#JWAS.misc.get_correlations"><code>JWAS.misc.get_correlations</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">get_correlations(samples_for_genetic_variances::Array{Array{Float64,2},1})</code></pre><ul><li>Get MCMC samples for correlations using MCMC samples for covariance matrces</li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/4.misc/src/using_marker_samples.jl#L159-L163">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="JWAS.misc.get_breeding_values" href="#JWAS.misc.get_breeding_values"><code>JWAS.misc.get_breeding_values</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">get_breeding_values(model)</code></pre><ul><li>Get esitimated breeding values and prediction error variances using samples of marker effects stored in <strong>files</strong>   for individuals defined by <code>outputEBV(model,IDs::Array{String,1})</code>, defaulting to all phenotyped individuals.</li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/4.misc/src/using_marker_samples.jl#L69-L74">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="JWAS.misc.QC" href="#JWAS.misc.QC"><code>JWAS.misc.QC</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">QC(infile,outfile;separator=&#39; &#39;,header=true,missing=false,MAF=0.1)</code></pre><ul><li>Quality control for input file <strong>infile</strong>, then write out to output file: <strong>outfile</strong>.<ul><li>Delete loci with minor allele frequency &lt; <strong>MAF</strong>.</li><li><strong>missing</strong> genotypes are replaced by column means.</li></ul></li><li>File format (header=true,separator=&#39;,&#39;,missing=9):</li></ul><pre><code class="language-none">Animal,marker1,marker2,marker3,marker4,marker5
S1,1,0,1,1,1
D1,2,0,9,2,1
O1,1,2,0,1,0
O3,0,0,2,1,1</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/reworkhow/JWAS.jl/blob/9eb03db9b320f14ec2facedc57081eac4ffc667e/src/4.misc/src/qc/qc.jl#L27-L41">source</a></section><footer><hr/><a class="previous" href="../workflow/"><span class="direction">Previous</span><span class="title">Workflow</span></a><a class="next" href="../internals/"><span class="direction">Next</span><span class="title">Internals</span></a></footer></article></body></html>
